// @aztec/circuits.js/dest/crs/index.js
import { createDebugLogger } from '@aztec/foundation/log';
import { fileURLToPath } from '@aztec/foundation/url';
import isNode from 'detect-node';
import { existsSync } from 'fs';
import { mkdir, open } from 'fs/promises';
import { dirname, join } from 'path';
/**
 * Downloader for CRS from the web or local.
 */
export class NetCrs {
    constructor(
    /**
     * The number of circuit gates.
     */
    numPoints) {
        this.numPoints = numPoints;
    }
    /**
     * Download the data.
     */
    async init() {
        // We need numPoints number of g1 points.
        const g1Start = 28;
        const g1End = g1Start + this.numPoints * 64 - 1;
        // Download required range of data.
        const response = await fetch('https://aztec-ignition.s3.amazonaws.com/MAIN%20IGNITION/monomial/transcript00.dat', {
            headers: {
                Range: `bytes=${g1Start}-${g1End}`,
            },
        });
        this.data = new Uint8Array(await response.arrayBuffer());
        await this.downloadG2Data();
    }
    /**
     * Download the G2 points data.
     */
    async downloadG2Data() {
        const g2Start = 28 + 5040001 * 64; // = 322_560_092
        const g2End = g2Start + 128 - 1;
        const response2 = await fetch('https://aztec-ignition.s3.amazonaws.com/MAIN%20IGNITION/monomial/transcript00.dat', {
            headers: {
                Range: `bytes=${g2Start}-${g2End}`,
            },
        });
        this.g2Data = new Uint8Array(await response2.arrayBuffer());
    }
    /**
     * Verification key data.
     * @returns The verification key.
     */
    getG1Data() {
        return this.data;
    }
    /**
     * G2 points data.
     * @returns The points data.
     */
    getG2Data() {
        return this.g2Data;
    }
}
/**
 * Downloader for CRS from a local file (for Node).
 */
export class FileCrs {
    constructor(
    /**
     * The number of circuit gates.
     */
    numPoints, path, offsetStart = true) {
        this.numPoints = numPoints;
        this.path = path;
        this.offsetStart = offsetStart;
    }
    /**
     * Read the data file.
     */
    async init() {
        // We need numPoints number of g1 points.
        const g1Start = this.offsetStart ? 28 : 0;
        const g1Length = this.numPoints * 64;
        const g2Start = 28 + 5040001 * 64; // = 322_560_092
        const g2Length = 128;
        // Lazily seek our data
        const fileHandle = await open(this.path, 'r');
        try {
            this.data = Buffer.alloc(g1Length);
            await fileHandle.read(this.data, 0, g1Length, g1Start);
            this.g2Data = Buffer.alloc(g2Length);
            await fileHandle.read(this.g2Data, 0, g2Length, this.offsetStart ? g2Start : g1Start + g1Length);
        }
        finally {
            await fileHandle.close();
        }
    }
    /**
     * Verification key data.
     * @returns The verification key.
     */
    getG1Data() {
        return this.data;
    }
    /**
     * G2 points data.
     * @returns The points data.
     */
    getG2Data() {
        return this.g2Data;
    }
}
/**
 * Generic CRS finder utility class.
 */
export class Crs {
    constructor(
    /**
     * The number of circuit gates.
     */
    numPoints, 
    /**
     * Option to save downloaded SRS on file.
     */
    saveOnFile = true) {
        this.numPoints = numPoints;
        this.saveOnFile = saveOnFile;
        this.logger = createDebugLogger('circuits:crs');
        /**
         * The path to our SRS object, assuming that we are in the aztec3-packages folder structure.
         */
        this.devPath = '/../../../../barretenberg/cpp/srs_db/ignition/monomial/transcript00.dat';
        /**
         * The path of our SRS object, if we downloaded on init.
         */
        this.localPath = `/../../resources/ignition/monomial/transcript00.dat`;
        if (isNode) {
            const devPath = join(fileURLToPath(import.meta.url), this.devPath);
            const localPath = join(dirname(fileURLToPath(import.meta.url)), this.localPath);
            const existsDev = existsSync(devPath);
            const existsLocal = existsSync(localPath);
            if (existsDev) {
                this.crs = new FileCrs(numPoints, devPath);
            }
            else if (existsLocal) {
                this.crs = new FileCrs(numPoints, localPath, false);
            }
            else {
                this.crs = new NetCrs(numPoints);
            }
        }
        else {
            this.crs = new NetCrs(numPoints);
        }
    }
    /**
     * Read CRS from our chosen source.
     */
    async init() {
        await this.crs.init();
        if (isNode) {
            const localPath = dirname(fileURLToPath(import.meta.url)) + this.localPath;
            const devPath = join(fileURLToPath(import.meta.url), this.devPath);
            // save downloaded CRS on file if we don't have a local or dev version
            if (this.saveOnFile && !existsSync(localPath) && !existsSync(devPath)) {
                const dir = localPath.substring(0, localPath.lastIndexOf('/'));
                await mkdir(dir, { recursive: true });
                const fileHandle = await open(localPath, 'w');
                const g1Data = Buffer.from(this.crs.getG1Data());
                try {
                    await fileHandle.write(g1Data);
                }
                catch (err) {
                    this.logger.warn('Failed to save CRS data: ', err.message);
                }
                const g2Data = Buffer.from(this.crs.getG2Data());
                try {
                    await fileHandle.write(g2Data, 0, g2Data.length, g1Data.length);
                    // appendFileSync(localPath, Buffer.from(g2Data));
                }
                catch (err) {
                    this.logger.warn('Failed to append G2 data: ', err.message);
                }
                finally {
                    await fileHandle.close();
                }
            }
        }
    }
    /**
     * Verification key data.
     * @returns The verification key.
     */
    getG1Data() {
        return this.crs.getG1Data();
    }
    /**
     * G2 points data.
     * @returns The points data.
     */
    getG2Data() {
        return this.crs.getG2Data();
    }
}
