// @aztec/circuits.js/dest/structs/global_variables.js
import { Fr } from '@aztec/foundation/fields';
import { BufferReader } from '@aztec/foundation/serialize';
import { serializeToBuffer } from '../utils/index.js';
/**
 * Global variables of the L2 block.
 */
export class GlobalVariables {
    constructor(
    /**
     * ChainId for the L2 block.
     */
    chainId, 
    /**
     * version for the L2 block.
     */
    version, 
    /**
     * Block number of the L2 block.
     */
    blockNumber, 
    /**
     * Timestamp of the L2 block.
     */
    timestamp) {
        this.chainId = chainId;
        this.version = version;
        this.blockNumber = blockNumber;
        this.timestamp = timestamp;
    }
    static from(fields) {
        return new GlobalVariables(...GlobalVariables.getFields(fields));
    }
    static empty() {
        return new GlobalVariables(Fr.zero(), Fr.zero(), Fr.zero(), Fr.zero());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new GlobalVariables(reader.readFr(), reader.readFr(), reader.readFr(), reader.readFr());
    }
    static fromJSON(obj) {
        return new GlobalVariables(Fr.fromString(obj.chainId), Fr.fromString(obj.version), Fr.fromString(obj.blockNumber), Fr.fromString(obj.timestamp));
    }
    static getFields(fields) {
        return [fields.chainId, fields.version, fields.blockNumber, fields.timestamp];
    }
    toBuffer() {
        return serializeToBuffer(...GlobalVariables.getFields(this));
    }
    toJSON() {
        return {
            chainId: this.chainId.toString(),
            version: this.version.toString(),
            blockNumber: this.blockNumber.toString(),
            timestamp: this.timestamp.toString(),
        };
    }
}
