// @aztec/circuits.js/dest/structs/private_circuit_public_inputs.js
import { Fr } from '@aztec/foundation/fields';
import { MAX_NEW_COMMITMENTS_PER_CALL, MAX_NEW_L2_TO_L1_MSGS_PER_CALL, MAX_NEW_NULLIFIERS_PER_CALL, MAX_PENDING_READ_REQUESTS_PER_CALL, MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, MAX_READ_REQUESTS_PER_CALL, NUM_FIELDS_PER_SHA256, RETURN_VALUES_LENGTH, } from '../cbind/constants.gen.js';
import { makeTuple } from '../utils/jsUtils.js';
import { serializeToBuffer } from '../utils/serialize.js';
import { CallContext } from './call_context.js';
import { HistoricBlockData } from './index.js';
import { ContractDeploymentData } from './tx_context.js';
/**
 * Public inputs to a private circuit.
 * @see abis/private_circuit_public_inputs.hpp.
 */
export class PrivateCircuitPublicInputs {
    constructor(
    /**
     * Context of the call corresponding to this private circuit execution.
     */
    callContext, 
    /**
     * Pedersen hash of function arguments.
     */
    argsHash, 
    /**
     * Return values of the corresponding function call.
     */
    returnValues, 
    /**
     * Read requests created by the corresponding function call.
     */
    readRequests, 
    /**
     * Pending read requests created by the corresponding function call.
     */
    pendingReadRequests, 
    /**
     * New commitments created by the corresponding function call.
     */
    newCommitments, 
    /**
     * New nullifiers created by the corresponding function call.
     */
    newNullifiers, 
    /**
     * The commitments those were nullified by the above newNullifiers.
     */
    nullifiedCommitments, 
    /**
     * Private call stack at the current kernel iteration.
     */
    privateCallStack, 
    /**
     * Public call stack at the current kernel iteration.
     */
    publicCallStack, 
    /**
     * New L2 to L1 messages created by the corresponding function call.
     */
    newL2ToL1Msgs, 
    /**
     * Hash of the encrypted logs emitted in this function call.
     * Note: Represented as an array of 2 fields in order to fit in all of the 256 bits of sha256 hash.
     */
    encryptedLogsHash, 
    /**
     * Hash of the unencrypted logs emitted in this function call.
     * Note: Represented as an array of 2 fields in order to fit in all of the 256 bits of sha256 hash.
     */
    unencryptedLogsHash, 
    /**
     * Length of the encrypted log preimages emitted in this function call.
     * Note: Here so that the gas cost of this request can be measured by circuits, without actually needing to feed
     *       in the variable-length data.
     */
    encryptedLogPreimagesLength, 
    /**
     * Length of the unencrypted log preimages emitted in this function call.
     */
    unencryptedLogPreimagesLength, 
    /**
     * Historic roots of the data trees, used to calculate the block hash the user is proving against.
     */
    historicBlockData, 
    /**
     * Deployment data of contracts being deployed in this kernel iteration.
     */
    contractDeploymentData, 
    /**
     * Chain Id of the instance.
     */
    chainId, 
    /**
     * Version of the instance.
     */
    version) {
        this.callContext = callContext;
        this.argsHash = argsHash;
        this.returnValues = returnValues;
        this.readRequests = readRequests;
        this.pendingReadRequests = pendingReadRequests;
        this.newCommitments = newCommitments;
        this.newNullifiers = newNullifiers;
        this.nullifiedCommitments = nullifiedCommitments;
        this.privateCallStack = privateCallStack;
        this.publicCallStack = publicCallStack;
        this.newL2ToL1Msgs = newL2ToL1Msgs;
        this.encryptedLogsHash = encryptedLogsHash;
        this.unencryptedLogsHash = unencryptedLogsHash;
        this.encryptedLogPreimagesLength = encryptedLogPreimagesLength;
        this.unencryptedLogPreimagesLength = unencryptedLogPreimagesLength;
        this.historicBlockData = historicBlockData;
        this.contractDeploymentData = contractDeploymentData;
        this.chainId = chainId;
        this.version = version;
    }
    /**
     * Create PrivateCircuitPublicInputs from a fields dictionary.
     * @param fields - The dictionary.
     * @returns A PrivateCircuitPublicInputs object.
     */
    static from(fields) {
        return new PrivateCircuitPublicInputs(...PrivateCircuitPublicInputs.getFields(fields));
    }
    /**
     * Create an empty PrivateCircuitPublicInputs.
     * @returns An empty PrivateCircuitPublicInputs object.
     */
    static empty() {
        return new PrivateCircuitPublicInputs(CallContext.empty(), Fr.ZERO, makeTuple(RETURN_VALUES_LENGTH, Fr.zero), makeTuple(MAX_READ_REQUESTS_PER_CALL, Fr.zero), makeTuple(MAX_PENDING_READ_REQUESTS_PER_CALL, Fr.zero), makeTuple(MAX_NEW_COMMITMENTS_PER_CALL, Fr.zero), makeTuple(MAX_NEW_NULLIFIERS_PER_CALL, Fr.zero), makeTuple(MAX_NEW_NULLIFIERS_PER_CALL, Fr.zero), makeTuple(MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL, Fr.zero), makeTuple(MAX_PUBLIC_CALL_STACK_LENGTH_PER_CALL, Fr.zero), makeTuple(MAX_NEW_L2_TO_L1_MSGS_PER_CALL, Fr.zero), makeTuple(NUM_FIELDS_PER_SHA256, Fr.zero), makeTuple(NUM_FIELDS_PER_SHA256, Fr.zero), Fr.ZERO, Fr.ZERO, HistoricBlockData.empty(), ContractDeploymentData.empty(), Fr.ZERO, Fr.ZERO);
    }
    /**
     * Serialize into a field array. Low-level utility.
     * @param fields - Object with fields.
     * @returns The array.
     */
    static getFields(fields) {
        return [
            // NOTE: Must have same order as CPP.
            fields.callContext,
            fields.argsHash,
            fields.returnValues,
            fields.readRequests,
            fields.pendingReadRequests,
            fields.newCommitments,
            fields.newNullifiers,
            fields.nullifiedCommitments,
            fields.privateCallStack,
            fields.publicCallStack,
            fields.newL2ToL1Msgs,
            fields.encryptedLogsHash,
            fields.unencryptedLogsHash,
            fields.encryptedLogPreimagesLength,
            fields.unencryptedLogPreimagesLength,
            fields.historicBlockData,
            fields.contractDeploymentData,
            fields.chainId,
            fields.version,
        ];
    }
    /**
     * Serialize this as a buffer.
     * @returns The buffer.
     */
    toBuffer() {
        return serializeToBuffer(...PrivateCircuitPublicInputs.getFields(this));
    }
}
