// @aztec/noir-contracts/src/artifacts/index.ts
// Auto generated module\n
import { ContractArtifact } from '@aztec/foundation/abi';
import BenchmarkingJson from './benchmarking_contract.json' assert { type: 'json' };
export const BenchmarkingContractArtifact = BenchmarkingJson as ContractArtifact;
import CardGameJson from './card_game_contract.json' assert { type: 'json' };
export const CardGameContractArtifact = CardGameJson as ContractArtifact;
import ChildJson from './child_contract.json' assert { type: 'json' };
export const ChildContractArtifact = ChildJson as ContractArtifact;
import DocsExampleJson from './docs_example_contract.json' assert { type: 'json' };
export const DocsExampleContractArtifact = DocsExampleJson as ContractArtifact;
import EasyPrivateTokenJson from './easy_private_token_contract.json' assert { type: 'json' };
export const EasyPrivateTokenContractArtifact = EasyPrivateTokenJson as ContractArtifact;
import EcdsaAccountJson from './ecdsa_account_contract.json' assert { type: 'json' };
export const EcdsaAccountContractArtifact = EcdsaAccountJson as ContractArtifact;
import EscrowJson from './escrow_contract.json' assert { type: 'json' };
export const EscrowContractArtifact = EscrowJson as ContractArtifact;
import ImportTestJson from './import_test_contract.json' assert { type: 'json' };
export const ImportTestContractArtifact = ImportTestJson as ContractArtifact;
import LendingJson from './lending_contract.json' assert { type: 'json' };
export const LendingContractArtifact = LendingJson as ContractArtifact;
import ParentJson from './parent_contract.json' assert { type: 'json' };
export const ParentContractArtifact = ParentJson as ContractArtifact;
import PendingCommitmentsJson from './pending_commitments_contract.json' assert { type: 'json' };
export const PendingCommitmentsContractArtifact = PendingCommitmentsJson as ContractArtifact;
import PriceFeedJson from './price_feed_contract.json' assert { type: 'json' };
export const PriceFeedContractArtifact = PriceFeedJson as ContractArtifact;
import SchnorrAccountJson from './schnorr_account_contract.json' assert { type: 'json' };
export const SchnorrAccountContractArtifact = SchnorrAccountJson as ContractArtifact;
import SchnorrHardcodedAccountJson from './schnorr_hardcoded_account_contract.json' assert { type: 'json' };
export const SchnorrHardcodedAccountContractArtifact = SchnorrHardcodedAccountJson as ContractArtifact;
import SchnorrSingleKeyAccountJson from './schnorr_single_key_account_contract.json' assert { type: 'json' };
export const SchnorrSingleKeyAccountContractArtifact = SchnorrSingleKeyAccountJson as ContractArtifact;
import StatefulTestJson from './stateful_test_contract.json' assert { type: 'json' };
export const StatefulTestContractArtifact = StatefulTestJson as ContractArtifact;
import TestJson from './test_contract.json' assert { type: 'json' };
export const TestContractArtifact = TestJson as ContractArtifact;
import TokenBridgeJson from './token_bridge_contract.json' assert { type: 'json' };
export const TokenBridgeContractArtifact = TokenBridgeJson as ContractArtifact;
import TokenJson from './token_contract.json' assert { type: 'json' };
export const TokenContractArtifact = TokenJson as ContractArtifact;
import UniswapJson from './uniswap_contract.json' assert { type: 'json' };
export const UniswapContractArtifact = UniswapJson as ContractArtifact;
