// @aztec/noir-contracts/src/types/child.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import ChildContractArtifactJson from '../artifacts/child_contract.json' assert { type: 'json' };

export const ChildContractArtifact = ChildContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract Child;
 */
export class ChildContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, ChildContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, ChildContract.artifact, wallet) as Promise<ChildContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ChildContract>(Point.ZERO, wallet, ChildContractArtifact, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<ChildContract>(publicKey, wallet, ChildContractArtifact, Array.from(arguments).slice(2));
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ChildContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** check_sender() */
    check_sender: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubGetValue(base_value: field) */
    pubGetValue: ((base_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubIncValue(new_value: field) */
    pubIncValue: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubIncValueInternal(new_value: field) */
    pubIncValueInternal: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubSetValue(new_value: field) */
    pubSetValue: ((new_value: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** setValueTwiceWithNestedFirst() */
    setValueTwiceWithNestedFirst: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** setValueTwiceWithNestedLast() */
    setValueTwiceWithNestedLast: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** value(input: field) */
    value: ((input: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** valueInternal(input: field) */
    valueInternal: ((input: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
