// @aztec/noir-contracts/src/types/import_test.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import ImportTestContractArtifactJson from '../artifacts/import_test_contract.json' assert { type: 'json' };

export const ImportTestContractArtifact = ImportTestContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract ImportTest;
 */
export class ImportTestContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, ImportTestContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, ImportTestContract.artifact, wallet) as Promise<ImportTestContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ImportTestContract>(
      Point.ZERO,
      wallet,
      ImportTestContractArtifact,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<ImportTestContract>(
      publicKey,
      wallet,
      ImportTestContractArtifact,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ImportTestContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** callNoArgs(target: field) */
    callNoArgs: ((target: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** callOpenFn(target: field) */
    callOpenFn: ((target: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** main(target: field) */
    main: ((target: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** pubCallOpenFn(target: field) */
    pubCallOpenFn: ((target: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
