// @aztec/noir-contracts/src/types/lending.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import LendingContractArtifactJson from '../artifacts/lending_contract.json' assert { type: 'json' };

export const LendingContractArtifact = LendingContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract Lending;
 */
export class LendingContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, LendingContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, LendingContract.artifact, wallet) as Promise<LendingContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<LendingContract>(
      Point.ZERO,
      wallet,
      LendingContractArtifact,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<LendingContract>(
      publicKey,
      wallet,
      LendingContractArtifact,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return LendingContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** _borrow(owner: field, to: field, amount: field) */
    _borrow: ((owner: FieldLike, to: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** _deposit(owner: field, amount: field, collateral_asset: field) */
    _deposit: ((owner: FieldLike, amount: FieldLike, collateral_asset: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** _repay(owner: field, amount: field, stable_coin: field) */
    _repay: ((owner: FieldLike, amount: FieldLike, stable_coin: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** _withdraw(owner: field, recipient: field, amount: field) */
    _withdraw: ((owner: FieldLike, recipient: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** borrow_private(secret: field, to: field, amount: field) */
    borrow_private: ((secret: FieldLike, to: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** borrow_public(to: field, amount: field) */
    borrow_public: ((to: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** deposit_private(from: field, amount: field, nonce: field, secret: field, on_behalf_of: field, collateral_asset: field) */
    deposit_private: ((
      from: FieldLike,
      amount: FieldLike,
      nonce: FieldLike,
      secret: FieldLike,
      on_behalf_of: FieldLike,
      collateral_asset: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** deposit_public(amount: field, nonce: field, on_behalf_of: field, collateral_asset: field) */
    deposit_public: ((
      amount: FieldLike,
      nonce: FieldLike,
      on_behalf_of: FieldLike,
      collateral_asset: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** get_asset(assetId: field) */
    get_asset: ((assetId: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_assets() */
    get_assets: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_position(owner: field) */
    get_position: ((owner: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** init(oracle_address: field, loan_to_value: field, collateral_asset: field, stable_coin: field) */
    init: ((
      oracle_address: FieldLike,
      loan_to_value: FieldLike,
      collateral_asset: FieldLike,
      stable_coin: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** repay_private(from: field, amount: field, nonce: field, secret: field, on_behalf_of: field, stable_coin: field) */
    repay_private: ((
      from: FieldLike,
      amount: FieldLike,
      nonce: FieldLike,
      secret: FieldLike,
      on_behalf_of: FieldLike,
      stable_coin: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** repay_public(amount: field, nonce: field, owner: field, stable_coin: field) */
    repay_public: ((
      amount: FieldLike,
      nonce: FieldLike,
      owner: FieldLike,
      stable_coin: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** update_accumulator() */
    update_accumulator: (() => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** withdraw_private(secret: field, to: field, amount: field) */
    withdraw_private: ((secret: FieldLike, to: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** withdraw_public(to: field, amount: field) */
    withdraw_public: ((to: FieldLike, amount: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
