// @aztec/noir-contracts/src/types/parent.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import ParentContractArtifactJson from '../artifacts/parent_contract.json' assert { type: 'json' };

export const ParentContractArtifact = ParentContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract Parent;
 */
export class ParentContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, ParentContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, ParentContract.artifact, wallet) as Promise<ParentContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<ParentContract>(Point.ZERO, wallet, ParentContractArtifact, Array.from(arguments).slice(1));
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<ParentContract>(publicKey, wallet, ParentContractArtifact, Array.from(arguments).slice(2));
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return ParentContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** enqueueCallsToChildWithNestedFirst(targetContract: field, targetSelector: field) */
    enqueueCallsToChildWithNestedFirst: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueueCallsToChildWithNestedLast(targetContract: field, targetSelector: field) */
    enqueueCallsToChildWithNestedLast: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueueCallsToPubEntryPoint(targetContract: field, targetSelector: field, targetValue: field) */
    enqueueCallsToPubEntryPoint: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      targetValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueueCallToChild(targetContract: field, targetSelector: field, targetValue: field) */
    enqueueCallToChild: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      targetValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueueCallToChildTwice(targetContract: field, targetSelector: field, targetValue: field) */
    enqueueCallToChildTwice: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      targetValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** enqueueCallToPubEntryPoint(targetContract: field, targetSelector: field, targetValue: field) */
    enqueueCallToPubEntryPoint: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      targetValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** entryPoint(targetContract: field, targetSelector: field) */
    entryPoint: ((targetContract: FieldLike, targetSelector: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** pubEntryPoint(targetContract: field, targetSelector: field, initValue: field) */
    pubEntryPoint: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      initValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** pubEntryPointTwice(targetContract: field, targetSelector: field, initValue: field) */
    pubEntryPointTwice: ((
      targetContract: FieldLike,
      targetSelector: FieldLike,
      initValue: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
