// @aztec/noir-contracts/src/types/pending_commitments.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import PendingCommitmentsContractArtifactJson from '../artifacts/pending_commitments_contract.json' assert { type: 'json' };

export const PendingCommitmentsContractArtifact = PendingCommitmentsContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract PendingCommitments;
 */
export class PendingCommitmentsContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, PendingCommitmentsContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, PendingCommitmentsContract.artifact, wallet) as Promise<PendingCommitmentsContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<PendingCommitmentsContract>(
      Point.ZERO,
      wallet,
      PendingCommitmentsContractArtifact,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<PendingCommitmentsContract>(
      publicKey,
      wallet,
      PendingCommitmentsContractArtifact,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return PendingCommitmentsContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** compute_note_hash_and_nullifier(contract_address: field, nonce: field, storage_slot: field, preimage: array) */
    compute_note_hash_and_nullifier: ((
      contract_address: FieldLike,
      nonce: FieldLike,
      storage_slot: FieldLike,
      preimage: FieldLike[],
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** dummy(amount: field, owner: field) */
    dummy: ((amount: FieldLike, owner: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_note_zero_balance(owner: field) */
    get_note_zero_balance: ((owner: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_then_nullify_note(expected_value: field, owner: field) */
    get_then_nullify_note: ((expected_value: FieldLike, owner: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** insert_note(amount: field, owner: field) */
    insert_note: ((amount: FieldLike, owner: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_bad_get_then_insert_flat(amount: field, owner: field) */
    test_bad_get_then_insert_flat: ((amount: FieldLike, owner: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert_then_get_then_nullify_all_in_nested_calls(amount: field, owner: field, insert_fn_selector: field, get_then_nullify_fn_selector: field, get_note_zero_fn_selector: field) */
    test_insert_then_get_then_nullify_all_in_nested_calls: ((
      amount: FieldLike,
      owner: FieldLike,
      insert_fn_selector: FieldLike,
      get_then_nullify_fn_selector: FieldLike,
      get_note_zero_fn_selector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert_then_get_then_nullify_flat(amount: field, owner: field) */
    test_insert_then_get_then_nullify_flat: ((amount: FieldLike, owner: FieldLike) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert1_then_get2_then_nullify2_all_in_nested_calls(amount: field, owner: field, insert_fn_selector: field, get_then_nullify_fn_selector: field, get_note_zero_fn_selector: field) */
    test_insert1_then_get2_then_nullify2_all_in_nested_calls: ((
      amount: FieldLike,
      owner: FieldLike,
      insert_fn_selector: FieldLike,
      get_then_nullify_fn_selector: FieldLike,
      get_note_zero_fn_selector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert2_then_get2_then_nullify1_all_in_nested_calls(amount: field, owner: field, insert_fn_selector: field, get_then_nullify_fn_selector: field) */
    test_insert2_then_get2_then_nullify1_all_in_nested_calls: ((
      amount: FieldLike,
      owner: FieldLike,
      insert_fn_selector: FieldLike,
      get_then_nullify_fn_selector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** test_insert2_then_get2_then_nullify2_all_in_nested_calls(amount: field, owner: field, insert_fn_selector: field, get_then_nullify_fn_selector: field) */
    test_insert2_then_get2_then_nullify2_all_in_nested_calls: ((
      amount: FieldLike,
      owner: FieldLike,
      insert_fn_selector: FieldLike,
      get_then_nullify_fn_selector: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
