// @aztec/noir-contracts/src/types/price_feed.ts
/* Autogenerated file, do not edit! */

/* eslint-disable */
import {
  AztecAddress,
  AztecAddressLike,
  CompleteAddress,
  Contract,
  ContractArtifact,
  ContractBase,
  ContractFunctionInteraction,
  ContractMethod,
  DeployMethod,
  EthAddress,
  EthAddressLike,
  FieldLike,
  Fr,
  Point,
  PublicKey,
  Wallet,
} from '@aztec/aztec.js';

import PriceFeedContractArtifactJson from '../artifacts/price_feed_contract.json' assert { type: 'json' };

export const PriceFeedContractArtifact = PriceFeedContractArtifactJson as ContractArtifact;

/**
 * Type-safe interface for contract PriceFeed;
 */
export class PriceFeedContract extends ContractBase {
  private constructor(completeAddress: CompleteAddress, wallet: Wallet, portalContract = EthAddress.ZERO) {
    super(completeAddress, PriceFeedContractArtifact, wallet, portalContract);
  }

  /**
   * Creates a contract instance.
   * @param address - The deployed contract's address.
   * @param wallet - The wallet to use when interacting with the contract.
   * @returns A promise that resolves to a new Contract instance.
   */
  public static async at(address: AztecAddress, wallet: Wallet) {
    return Contract.at(address, PriceFeedContract.artifact, wallet) as Promise<PriceFeedContract>;
  }

  /**
   * Creates a tx to deploy a new instance of this contract.
   */
  public static deploy(wallet: Wallet) {
    return new DeployMethod<PriceFeedContract>(
      Point.ZERO,
      wallet,
      PriceFeedContractArtifact,
      Array.from(arguments).slice(1),
    );
  }

  /**
   * Creates a tx to deploy a new instance of this contract using the specified public key to derive the address.
   */
  public static deployWithPublicKey(publicKey: PublicKey, wallet: Wallet) {
    return new DeployMethod<PriceFeedContract>(
      publicKey,
      wallet,
      PriceFeedContractArtifact,
      Array.from(arguments).slice(2),
    );
  }

  /**
   * Returns this contract's artifact.
   */
  public static get artifact(): ContractArtifact {
    return PriceFeedContractArtifact;
  }

  /** Type-safe wrappers for the public methods exposed by the contract. */
  public methods!: {
    /** fetch_price(assetId: field) */
    fetch_price: ((assetId: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** get_price(asset_id: field) */
    get_price: ((asset_id: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** set_price(asset_id: field, price: integer) */
    set_price: ((asset_id: FieldLike, price: bigint | number) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;
  };
}
