// @aztec/circuits.js/dest/cbind/types.js
// Type mappings for cbinds
// Can either export things directly or handle
// naming differences with the 'as' syntax
import { toBufferBE } from '@aztec/foundation/bigint-buffer';
import { MembershipWitness } from '../structs/membership_witness.js';
/**
 * Alias for msgpack which expects a MembershipWitness + N name.
 */
export class MembershipWitness20 extends MembershipWitness {
    constructor(
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath) {
        super(20, leafIndex.toBigInt(), siblingPath);
    }
}
/**
 * Alias for msgpack which expects a MembershipWitness + N name.
 */
export class MembershipWitness16 extends MembershipWitness {
    constructor(
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath) {
        super(16, leafIndex.toBigInt(), siblingPath);
    }
}
/**
 * Alias for msgpack which expects a MembershipWitness + N name.
 */
export class MembershipWitness8 extends MembershipWitness {
    constructor(
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath) {
        super(8, leafIndex.toBigInt(), siblingPath);
    }
}
/**
 * Alias for msgpack which expects a MembershipWitness + N name.
 */
export class MembershipWitness4 extends MembershipWitness {
    constructor(
    /**
     * Index of a leaf in the Merkle tree.
     */
    leafIndex, 
    /**
     * Sibling path of the leaf in the Merkle tree.
     */
    siblingPath) {
        super(4, leafIndex.toBigInt(), siblingPath);
    }
}
/**
 * Coerce a variety of types to a buffer.
 * Makes msgpack output easier to manage as this can handle a few cases.
 * @param bufferable - The value to coerce.
 */
export function toBuffer(bufferable) {
    if (typeof bufferable === 'bigint') {
        return toBufferBE(bufferable, 32);
    }
    else if (bufferable instanceof Buffer) {
        return bufferable;
    }
    else {
        return bufferable.toBuffer();
    }
}
export { AggregationObject as NativeAggregationState, AztecAddress as Address, VerificationKey as VerificationKeyData, Fr, Fq, G1AffineElement, NewContractData, FunctionData, OptionallyRevealedData, PublicDataRead, PublicDataUpdateRequest, ReadRequestMembershipWitness, CombinedAccumulatedData, FinalAccumulatedData, HistoricBlockData, ContractDeploymentData, TxContext, CombinedConstantData, CompleteAddress, KernelCircuitPublicInputs, KernelCircuitPublicInputsFinal, Proof, PreviousKernelData, CallContext, ContractStorageUpdateRequest, ContractStorageRead, PublicCircuitPublicInputs, PublicCallStackItem, PublicCallData, PublicKernelInputs, CircuitError, Point, Coordinate, GlobalVariables, PrivateKernelInputsOrdering, PrivateCallData, PrivateCallStackItem, PrivateCircuitPublicInputs, PrivateKernelInputsInit, PrivateKernelInputsInner, TxRequest, PreviousRollupData, AppendOnlyTreeSnapshot, BaseRollupInputs, NullifierLeafPreimage, BaseOrMergeRollupPublicInputs, ConstantRollupData, MergeRollupInputs, RootRollupInputs, RootRollupPublicInputs, } from '../structs/index.js';
export { FunctionSelector } from '@aztec/foundation/abi';
